From 020f50823c7eccd543ed3f6fd769b23cd817f1b8 Mon Sep 17 00:00:00 2001
From: Sundar Iyer <sundar.iyer@intel.com>
Date: Thu, 1 Aug 2013 11:33:21 +0530
Subject: [PATCH] [PORT from k3.4 to k3.10] acpi/processor_perflib: Enable SW
 coordination

BZ: 144101

A multi-module Atom Silvermont implementation currently doesnt support
HW coordination across all modules. Use the SW coordination instead:
this is easier and quicker.

K3.4 BZ: 127510

Reported-by: Yongnian Le <yongnian.le@intel.com>
Signed-off-by: Sundar Iyer <sundar.iyer@intel.com>
Signed-off-by: Stephane Gasparini <stephane.gasparini@intel.com>
Signed-off-by: Vishwanath Somayaji <vishwanath.somayaji@intel.com>

Orig-Change-Id: I0bbbe2ac193d0cf477aff19777d7c252e2cafa51
Change-Id: I94984870acd5c34491e838a10b4925de8064759c
Reviewed-on: http://android.intel.com:8080/137538
Reviewed-by: Somayaji, Vishwanath <vishwanath.somayaji@intel.com>
Reviewed-by: Mansoor, Illyas <illyas.mansoor@intel.com>
Reviewed-by: Kasagar, Srinidhi <srinidhi.kasagar@intel.com>
Reviewed-by: Englund, Hakan <hakan.englund@intel.com>
Tested-by: Delance, OlivierX <olivierx.delance@intel.com>
Reviewed-by: Yang, Fei <fei.yang@intel.com>
Reviewed-by: cactus <cactus@intel.com>
Tested-by: cactus <cactus@intel.com>
---
 drivers/acpi/processor_perflib.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/acpi/processor_perflib.c b/drivers/acpi/processor_perflib.c
--- a/drivers/acpi/processor_perflib.c	2015-01-29 12:26:36.012718152 -0800
+++ b/drivers/acpi/processor_perflib.c	2015-01-29 12:26:41.960718348 -0800
@@ -37,6 +37,8 @@
 #include <asm/cpufeature.h>
 #endif
 
+#define X86_ATOM_ARCH_SLM              (0x37)
+
 #define PREFIX "ACPI: "
 
 #define ACPI_PROCESSOR_CLASS		"processor"
@@ -659,7 +661,7 @@ int acpi_processor_preregister_performan
 		goto err_ret;
 
 	/*
-	 * Now that we have _PSD data from all CPUs, lets setup P-state 
+	 * Now that we have _PSD data from all CPUs, lets setup P-state
 	 * domain info.
 	 */
 	for_each_possible_cpu(i) {
@@ -685,6 +687,14 @@ int acpi_processor_preregister_performan
 		else if (pdomain->coord_type == DOMAIN_COORD_TYPE_SW_ANY)
 			pr->performance->shared_type = CPUFREQ_SHARED_TYPE_ANY;
 
+		/* hard code the cpufreq policy to ensure that SW coordination
+		 * is in place; on 2CMP SLM systems, the HW cannot arbitrate and
+		 * hence SW coordination is neccessary!
+		*/
+		if (boot_cpu_data.x86_model == X86_ATOM_ARCH_SLM)
+			pr->performance->shared_type = CPUFREQ_SHARED_TYPE_ALL;
+
+
 		for_each_possible_cpu(j) {
 			if (i == j)
 				continue;
@@ -725,7 +735,7 @@ int acpi_processor_preregister_performan
 			if (match_pdomain->domain != pdomain->domain)
 				continue;
 
-			match_pr->performance->shared_type = 
+			match_pr->performance->shared_type =
 					pr->performance->shared_type;
 			cpumask_copy(match_pr->performance->shared_cpu_map,
 				     pr->performance->shared_cpu_map);
